components/renku_data_services/notebooks/config/__init__.py

```diff
@@ -236,6 +221,99 @@ def from_env(cls, db_config: DBConfig) -> Self:
             git_provider_helper=git_provider_helper,
             k8s_client=k8s_client,
             k8s_v2_client=k8s_v2_client,
-            k8s_cached_client=k8s_cached_client,
+            k8s_db_cache=k8s_db_cache,
             _kr8s_api=kr8s_api,
         )
+
+
+class _KubeConfig:
+    def __init__(
+        self,
+        kubeconfig: str | None = None,
+        current_context_name: str | None = None,
+        ns: str | None = None,
+        sc: str | None = None,
+        url: str | None = None,
+    ) -> None:
+        self._kubeconfig = kubeconfig
+        self._ns = ns
+        self._current_context_name = current_context_name
+        self._sc = sc
+        self._url = url
+
+    def _sync_api(self) -> Union[kr8s.Api, kr8s._AsyncApi]:
+        return kr8s.api(
+            kubeconfig=self._kubeconfig,
+            namespace=self._ns,
+            context=self._current_context_name,
+        )
+
+    def _async_api(self) -> kr8s.asyncio.Api:
+        """Create an async api client from sync code.
+
+        Kr8s cannot return an AsyncAPI instance from sync code, and we can't easily make all our config code async,
+        so this method is a direct copy of the kr8s sync client code, just that it returns an async client.
+        """
+        ret = kr8s._async_utils.run_sync(kr8s.asyncio.api)(
+            url=self._url,
+            kubeconfig=self._kubeconfig,
+            serviceaccount=self._sc,
+            namespace=self._ns,
+            context=self._current_context_name,
+            _asyncio=True,  # This is the only line that is different from kr8s code
+        )
+        assert isinstance(ret, kr8s.asyncio.Api)
+        return ret
+
+    def api(self, _async: bool = True) -> Union[kr8s.Api, kr8s._AsyncApi]:
+        # Instantiate the Kr8s Api object based on the configuration
+        if _async:
+            return self._async_api()
+        else:
+            return self._sync_api()
+
+
+class _KubeConfigEnv(_KubeConfig):
+    def __init__(self) -> None:
+        super().__init__(ns=os.environ.get("K8S_NAMESPACE", "default"))
+
+
+class _KubeConfigYaml(_KubeConfig):
+    def __init__(self, kubeconfig: str) -> None:
+        super().__init__(kubeconfig=kubeconfig)
+
+        with open(kubeconfig) as stream:
+            self._conf = yaml.safe_load(stream)
+
+        self._current_context_name = self._conf.get("current-context", None)
+        if self._current_context_name is not None:
+            for context in self._conf.get("contexts", []):
+                name = context.get("name", None)
+                inner = context.get("context", None)
+                if inner is not None and name is not None and name == self._current_context_name:
+                    self._ns = inner.get("namespace", None)
+                    break
```

Comment:
suggestion: There is no need for the contents of the kubeconfig to be part of the class. Since they will contain some pretty sensitive and powerful info it is better to not store on the class at all.


```suggestion
class _KubeConfigYaml(_KubeConfig):
    def __init__(self, kubeconfig: str) -> None:
        super().__init__(kubeconfig=kubeconfig)

        with open(kubeconfig) as stream:
            _conf = yaml.safe_load(stream)

        self._current_context_name = _conf.get("current-context", None)
        if self._current_context_name is not None:
            for context in _conf.get("contexts", []):
                name = context.get("name", None)
                inner = context.get("context", None)
                if inner is not None and name is not None and name == self._current_context_name:
                    self._ns = inner.get("namespace", None)
                    break
```

Reply:
done